torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -2610,7 +2663,38 @@ void ProcessGroupNCCL::groupStart() {
 
 void ProcessGroupNCCL::groupEnd() {
 #if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+#ifndef NCCL_HAS_COMM_NONBLOCKING
+  C10D_NCCL_CHECK(ncclGroupEnd(), c10::nullopt);
+#else
+  if (!nccl_use_nonblocking()) {
+    C10D_NCCL_CHECK(ncclGroupEnd(), c10::nullopt);
+  } else {
+    TORCH_WARN(
+        "ProcessGroupNCCL::groupEnd() called in nonblocking communicator mode without involved communicators specified; gathering all mapped communicators...");
+    std::unique_lock<std::mutex> lock(mutex_);
+    std::vector<std::shared_ptr<NCCLComm>> ncclComms_;
+    for (auto& it : devNCCLCommMap_) {
+      ncclComms_.insert(ncclComms_.end(), it.second.begin(), it.second.end());
+    }
```

Comment:
Hmm, this could be a bit dangerous as `devNCCLCommMap_` is a cache for all the communicators of all the devices under this process, which is a bigger scope than the communicators involved in the group operation.

I think we should assert an internal error here, as `groupEnd()` is only called under this umbrella as in the `endCoalescing()` code:
```
if (!nccl_use_nonblocking()) {
    groupEnd();
  }
```